Add support for (very) large-scale data

jorainer opened the pull request:
This PR adds a new *xcms* result object `XcmsExperimentHdf5` that stores all preprocessing results to a local file in HDF5 format. Thus, no data/results are kept in memory reducing the memory demand to a minimum.

For an example and performance evaluation of this backend check the *large_scale* branch on Metabonaut: https://github.com/rformassspectrometry/Metabonaut/tree/large_scale_data

There are also other, smaller, changes in this PR including:

- improve performance of `chromatogram` call by reading only spectra within the retention time ranges
- linking to Metabonaut
- update **all** documentations to roxygen/markdown. Fix links in documentation.

Apologies for the very large PR - there are unit tests for all functions, so the code should be OK. So maybe just discuss/evaluate the concept?

jorainer committed:
Add initial ideas and code

jorainer committed:
feat: add some more functions related to HDF5 files.

jorainer committed:
feat: add additional functionaliy related to HDF5 files

jorainer committed:
feat: add XcmsExperimentHdf class and first methods

jorainer committed:
fix: small fixes.

jorainer committed:
feat: add refineChromPeaks,XcmsExperimentHdf5

jorainer committed:
feat: add .h5_chrom_peaks function

jorainer committed:
feat: add chromPeaks,XcmsExperimentHdf5 method

jorainer committed:
feat: add parameter columns to chromPeaks for XcmsExperiment

jorainer committed:
feat: add correspondence functionality

jorainer committed:
feat: add featureValues,XcmsExperimentHdf5

jorainer committed:
feat: add functionality to get chrom peaks for features

jorainer committed:
Merge branch 'devel' into large_scale

jorainer committed:
Add functionality related to retention time alignment

jorainer committed:
feat: add adjustRtime and related functions for XcmsExperimentHdf5

jorainer committed:
Small fixes and improvements

jorainer committed:
Add chromPeakData,XcmsExperimentHdf5 function

jorainer committed:
Performance improvement for chromPeakData,XcmsExperimentHdf5

jorainer committed:
feat: add chromatogram,XcmsExperimentHdf5

jorainer committed:
tests: fix unit test

jorainer committed:
Add first gap filling functionality

jorainer committed:
feat: add featureArea for XcmsExperimentHdf5

jorainer committed:
tests: fix unit tests

jorainer committed:
Merge remote-tracking branch 'origin/devel' into large_scale

jorainer committed:
feat: add fillChromPeaks,XcmsExperimentHdf5

jorainer committed:
feat: add dropFilledChromPeaks,XcmsExperimentHdf5

jorainer committed:
feat: add filterMsLevel,XcmsExperimentHdf5

jorainer committed:
feat: add filterRt,XcmsExperimentHdf5

jorainer committed:
feat: add filterIsolationWindow,XcmsExperimentHdf5

jorainer committed:
feat: add functions to convert between xcms result objects

jorainer committed:
Merge branch 'devel' into large_scale

jorainer committed:
feat: add support for additional functions

- Add support for `dropChromPeaks()`.
- Add support for `dropFeatureDefinitions()`.

jorainer committed:
feat: add missing refineChromPeaks for XcmsExperimentHdf5

jorainer committed:
feat: add chromPeakSpectra,XcmsExperimentHdf5

- Add the `chromPeakSpectra()` method for `XcmsExperimentHdf5` objects.

jorainer committed:
feat: add chromPeakSpectra and featureSpectra for XcmsExperimentHdf5

jorainer committed:
docs: update to roxygen

jorainer committed:
Merge branch 'devel' into large_scale

jorainer committed:
docs: convert documentation to markdown format

jorainer committed:
Merge branch 'jomain' into large_scale

jorainer committed:
feat: add featureChromatograms for XcmsExperimentHdf5

- Add `featureChromatograms()` for `XcmsExperimentHdf5`.
- Tune chromatogram extraction methods.

jorainer committed:
feat: add filterChromPeaks,XcmsExperimentHdf5

jorainer committed:
Merge branch 'AnnotatedDataFrame' into large_scale

jorainer committed:
feat: add filterFeatureDefinitions,XcmsExperimentHdf5

jorainer committed:
docs: update documentation to markdown

jorainer committed:
Merge branch 'devel' into large_scale

jorainer committed:
feat: add manualChromPeaks,XcmsExperimentHdf5 method

jorainer committed:
refactor: use reduce from MsCoreUtils

jorainer committed:
fix: smaller fixes

jorainer committed:
feat: add chromPeakSummary,XcmsExperimentHdf5

jwokaty committed:
bump x.y.z version to odd y following creation of RELEASE_3_21 branch

jwokaty committed:
bump x.y.z version to even y prior to creation of RELEASE_3_21 branch

jorainer committed:
docs: restructure documentation and move aliases to hidden aliases

jorainer committed:
docs: small fixes to the vignettes

jorainer committed:
docs: fix aliases and add Metabonaut links to vignettes

jorainer committed:
docs: update NEWS

jorainer committed:
docs: small fix in vignette

jorainer committed:
Merge branch 'devel' into large_scale

jorainer committed:
ci: update R version

jorainer committed:
fix: chromatograms() to throw error if NA ranges provided

- `chromatograms()` to throw an error if parameter `mz` or `rt` contain missing
  values.

jorainer committed:
ci: set environment variables to fix data.table install

jorainer committed:
ci: install gettext on macOS

jorainer committed:
fix: disable inclusion of libintl.h in massifquant

sneumann commented on NAMESPACE:
```diff
@@ -291,156 +294,6 @@ exportClasses(
     "FilterIntensityParam",
     "ChromPeakAreaParam"
 )
-## Param methods
```
Were those removed entirely ? I see a few new exports elsewhere, but only a minority of these here

jorainer replied:
I removed all of the *getter/setter* methods for the `*Param` classes (please let me know if there are any left - I would like to remove all).

Why I removed these:

- it clutters the documentation with seldomly, if at all, used functions
- users will almost never use these functions. It's *easier* to just create a new param object than changing the value of a previously created one.

sneumann commented on NEWS.md:
```diff
@@ -1,11 +1,29 @@
+
+## Changes in version 4.7.1
+
+- Change the naming convention of chromatographic peaks to include also the MS
```
That starts to put semantics into the identifier. I am extra careful with that, code needs to check MS level <=9. Are there any other semantics that could creep around the corner and need consideration ? Ion Mobility ? Probably not, remains an attribute of the peaks. Can you still rely on the leading digit in older datasets ? CP2001 could also be the 2001st peak in an old dataset ...

jorainer replied:
The chrom peak identifiers were always just thought as an identifier **within** the same data set. So, IMHO, it does not matter what values these identifiers have (whether it's a running number or even just a random number, as long as it is unique). I agree that adding semantics to identifiers is not ideal - but for the new `XcmsExperimentHdf5` I needed something that allows me to define chromatographic peaks IDs separately for each process/file. Peak detection is performed on a per sample (file) basis and each of  these processes stores the identified peaks to the HDF5 file without knowing anything about the other projects. That's why I came up with this solution - to still have **unique** chrom peak IDs within the same data set. I'm of course open to any other working solution :)

sneumann commented on NEWS.md:
```diff
@@ -1,11 +1,29 @@
+- Optimization and performance improvements for extraction of chromatographic
+  data. This includes using `MsCoreUtils::reduce()`.
+- Restructure and clean-up of documentation.
+- Don't export unnecessary get/set methods for `Param` classes.
```
Ok. Methods are not neccessary or their export ? Could we have any old users of those functions ?

jorainer replied:
Hard to tell if there are any users out there - I guess very few if any (we never really promoted the idea of using getter/setter methods of the parameter classes). I would change and see what happens - my guess is users come forward and will open issues, then we can address it.

sneumann commented on R/DataClasses.R:
```diff
@@ -638,112 +627,90 @@ setClass("CentWaveParam",
 #' Gary Siuzdak. "XCMS: Processing Mass Spectrometry Data for Metabolite
 #' Profiling Using Nonlinear Peak Alignment, Matching, and Identification"
-#' \emph{Anal. Chem.} 2006, 78:779-787.
+#' *Anal. Chem.* 2006, 78:779-787.
```
Future citations could include the DOI.

jorainer replied:
good point. I'll add the DOI for all references.

jorainer replied:
I added the DOIs in commit https://github.com/sneumann/xcms/commit/c120f61f427c6cce120aa8a9e9225550cb98082b

sneumann commented on R/DataClasses.R:
```diff
@@ -1204,73 +1139,51 @@ setClass("MSWParam",
+#' p <- CentWavePredIsoParam(maxCharge = 4, snthresh = 25)
 #' p
-#'
+NULL
```
I'd love to learn what these two NULL do, and what's the difference between 'em.

jorainer replied:
The `NULL` after a roxygen documention just ensures that the documentation file (Rd) is created - without adding also the next function to it. For general documentation or collection/topics of function I use that, i.e., write the roxygen documentation and add a `NULL` at the end. Then, I *add* all related functions/methods/classes to the documentation using `#' @rdname <name of the general, topic, documentation>`

sneumann commented on R/DataClasses.R:
```diff
@@ -1660,185 +1575,164 @@ setClass("MsFeatureData", contains = c("environment"),
+#'   with feature definitions and phenodata information, into a
+#'   [SummarizedExperiment::SummarizedExperiment()]. See help page for details.
 #'
-#' \item \code{\link{quantify}} extract feature intensities and put them, along
```
Is there some kind of deprecation documentation ? Or porting guide to xcms-4.7.x ?

jorainer replied:
I'm a bit confused what you mean. If this refers to the documentation, I did not delete any documentation entry, the `quantify()` method is described a bit above. I did not change/remove any old functionality. This is just some changes in the documentation, where I change to the newer *md*-type format - and eventually re-order functions alphabetically or group them by topic into *sections*.

sneumann commented on R/DataClasses.R:
```diff
@@ -2186,4 +2083,8 @@ setClass("FilterIntensityParam",
+
+############################################################
+## Class unions
+setClassUnion("characterOrNULL", c("character", "NULL"))
```
Ah, I was wondering where they went :-)

jorainer replied:
yes, wanted to somehow group them together.

sneumann commented on R/MsExperiment-functions.R:
```diff
@@ -63,7 +63,8 @@
     x <- filterMsLevel(x, msLevel)
     pkd <- Spectra::peaksData(x, columns = c("mz", "intensity"),
-                              f = factor(), BPPARAM = SerialParam())
+                              f = factor(), return.type = "list",
```
This inserted argument could confuse calling code that does not use named paramters, but relies on position. Is there a check whether `return.type` is a `SerialParam`to avoid silent stupid errors ?

jorainer replied:
hm. good point. Or I move the new parameter `return.type` **after** `BPPARAM`?

jorainer replied:
After checking the code again - this does not apply. This function is just called internally - without any possibility of the user to provide or change the parameters to this call.

sneumann commented on R/XcmsExperiment-functions.R:
```diff
@@ -1155,15 +1158,21 @@ featureArea <- function(object, mzmin = min, mzmax = max, rtmin = min,
+    which(keep & .is_chrom_peaks_within_mz_rt(pks, rt, mz, ppm, type))
+}
+
+.is_chrom_peaks_within_mz_rt <- function(x, rt = numeric(), mz = numeric(),
```
Bikeshedding: `.is_chrom_peak_within_mz_rt` singular ? Whatever, but consistent with names elsewhere.

jorainer replied:
yes, agree - it sounds terrible :D . I'll change.

jorainer replied:
Done in https://github.com/sneumann/xcms/commit/c120f61f427c6cce120aa8a9e9225550cb98082b

sneumann commented on R/XcmsExperiment-functions.R:
```diff
@@ -1173,20 +1182,23 @@ featureArea <- function(object, mzmin = min, mzmax = max, rtmin = min,
             mz[2L] <- mz[2L] + mz[2L] * ppm / 1e6
         if (type == "any")
-            keep <- keep & pks[, "mzmin"] <= mz[2L] & pks[, "mzmax"] >= mz[1L]
+            keep <- keep & x[, "mzmin"] <= mz[2L] & x[, "mzmax"] >= mz[1L]
```
I have difficulties finding `x`, and what happened to pks

jorainer replied:
`x` is the input parameter to `.is_chrom_peaks_within_mz_rt()`. It was `pks` before, but I renamed it to `x` because having the *main* input parameter to a function called `x` is more consistent with *general* R usage (same as the main parameter to a method is called `object`). I was not that consistent in the past, but for newer functions I try to stick to this more.

sneumann commented on R/XcmsExperiment.R:
```diff
@@ -870,11 +887,24 @@ setMethod(
                  "supported", call. = FALSE)
+        if (length(hdf5File)) {
+            if (file.exists(hdf5File))
+                stop("File ", hdf5File, " already exists. Replacing results ",
```
If I have a vignette, does that mean to be on the safe side I need to `system("rm file.hdf5")` in the beginning ? Alternatively, there could be a `force.overwrite=TRUE` parameter somewhere

jorainer replied:
Thanks! that's a great idea! I added a `force.overwrite` parameter in commit https://github.com/sneumann/xcms/commit/c120f61f427c6cce120aa8a9e9225550cb98082b

sneumann commented on R/XcmsExperimentHdf5-functions.R:
```diff
@@ -0,0 +1,1666 @@
+##
+################################################################################
+
+## ## WE MIGHT ACTUALLY NOT NEED THIS!
```
Happy to learn why we may or may not need it

jorainer replied:
This refers to the commented-out function below (`.h5_feature_chrom_peaks_sample()`). I implemented it but did not use in the end - I just wanted to keep the code as example/reference.

sneumann commented on R/XcmsExperimentHdf5-functions.R:
```diff
@@ -0,0 +1,1666 @@
+#' Properties of the HDF5 file used for on-disk storage of xcms results:
```
Is there versioning info about the "xcms schema version" of the hdf5 in case we change it in the future, and can gracefully reject files in the old schema ? Or convert / migrate files ?

jorainer replied:
I did not define a schema version - unless you can convince me otherwise, I think that will result in a bit of a large code overhead - also knowing that this HDF5 format is really just thought for very large experiments and will only be used sparsely.

sneumann commented on R/XcmsExperimentHdf5-functions.R:
```diff
@@ -0,0 +1,1666 @@
+                  medret = c(value, "rt"),
+                  maxint = c(value, intensity))
+    if (!all(col %in% cn))
+        stop("Not all requested columns available. Please make sure 'value' ",
```
Debugging could be simplified if we print the missing colnames. 
Why could they be missing ? Missing processing step ? Or new "xcms schema version" ?

jorainer replied:
This is just fail safe - in case some (future?) other chrom peak detection algorithm is implemented that does not store these column names. I will add the supported column names.

jorainer replied:
added in https://github.com/sneumann/xcms/commit/de9ae22946e7f5226ad2e5eaa4c30c6ddc4ca417

sneumann commented on R/XcmsExperimentHdf5-functions.R:
```diff
@@ -0,0 +1,1666 @@
+#'
+#' Setting row and column names requires additional read steps and is hence
+#' considerably slower than *just* importing the data. Also, if possible,
+#' consider importing only the required column(s) using parameter `column`.
```
Bikeshedding again, could that be parameter `columns=c(...)` in plural ?

jorainer replied:
sure - I realized the parameter was actually called `name`, not `colname` - I will change to `names`.

jorainer replied:
Indeed. I just cross-checked. The downstream functions use parameter `columns` not `column`, so this was just a mis-spelling in the description. I'll update.

sneumann commented on R/XcmsExperimentHdf5.R:
```diff
@@ -0,0 +1,1501 @@
+#' The *xcms* result objects [XcmsExperiment()] and [XCMSnExp()] keep all
+#' preprocessing results in memory and can thus (depending on the size of the
+#' data set) require a large amount of memory. In contrast, the
+#' `XcmsExperimentHdf5` class, by using an on-disk data storage mechanism,
```
How would the hdf5 approach actually work in, say, a Galaxy setting ? Where some of the operations can be executed in parallel, and then some operations (e..g correspondance analysis) join that into a single file.

jorainer replied:
Parallel processing **must** be controlled by *xcms* (it's built in and can be configured with `BPPARAM` and `chunkSize` parameters). Data is processed in chunks of files, in parallel, results are then written, after processing one chunk, to the file. But that's similar to the *in-memory* approaches we have in *xcms*, where the result is collected at the end from the individual processes - only, here we don't do this at the very end of the processing (i.e. after processing all samples), but after processing the samples of one chunk.

sneumann commented on R/XcmsExperimentHdf5.R:
```diff
@@ -0,0 +1,1501 @@
+
+#' @rdname hidden_aliases
+setMethod("hasFilledChromPeaks", "XcmsExperimentHdf5", function(object) {
+    length(object@gap_peaks_ms_level) > 0
```
So filled peaks are counted in gap_peaks. And what ms_level is that ? ms1 ?

jorainer replied:
no, the `@gap_peaks_ms_level` contains just the MS levels in which gap-filled peaks are present. So, if gap filling was performed on MS1 then we would have `@gap_peaks_ms_level = c(1L)`

sneumann commented on R/XcmsExperimentHdf5.R:
```diff
@@ -0,0 +1,1501 @@
+                expandMz = expandMz, ppm = ppm, skipFilled = skipFilled,
+                peaks = peaks, chromPeakColumns = chromPeakColumns)
+        })
+        res <- Spectra:::.concatenate_spectra(res)
```
Should that be exported by Spectra in the future ?

jorainer replied:
I have to check the function - could be that it's already exported.

sneumann replied:
This one without leading dot ? 
https://github.com/rformassspectrometry/Spectra/blob/82f773a3a8a305341b9baed3216632599649719d/NAMESPACE#L14

jorainer replied:
yes. I will update in the next PR.

sneumann commented on R/functions-XCMSnExp.R:
```diff
@@ -577,35 +584,6 @@ dropGenericProcessHistory <- function(x, fun) {
 }
 
-#' @rdname adjustRtime
-adjustRtimePeakGroups <- function(object, param = PeakGroupsParam(),
```
This is entirely gone now ?

jorainer replied:
on the contrary - I converted that to a method - so I added dedicated implementations for `XcmsExperiment`, `XcmsExperimentHdf5` and `XCMSnExp`

sneumann commented on R/methods-xcmsFragments.R:
```diff
@@ -3,6 +3,7 @@
 ############################################################
 ## show
+#' @rdname hidden_aliases
 setMethod("show", "xcmsFragments", .xcmsFragments.show)
```
How useful is xcmsFragments nowadays ? Would this devel cycle be a good time to deprecate it and mark for removal in the next cycle ?

jorainer replied:
I would be OK with that. I don't know what this class actually is/was used for... But we should maybe start deprecation in a separate branch/PR.

sneumann commented on R/ramp.R:
```diff
@@ -1,228 +0,0 @@
-## # rampInit <- function() {
```
Rest in pieces :-)

sneumann commented on tests/testthat/test_XcmsExperimentHdf5-functions.R:
```diff
@@ -0,0 +1,1083 @@
+library(rhdf5)
```
should the library() be in the parent directory, for all *hdf5 tests  ?

jorainer replied:
I wanted to keep the HDF5 functionality as separate as possible thus I did not add it into the main *testthat.R* file. I find it *cleaner* if the unit test file specifically sets the stage for its R source file.

sneumann reviewed (commented):
Hi, awesome work. Some cosmetic questions here and there. Yours, Steffen

jorainer committed:
address Steffen's comments

- Add DOI for all references.
- Add parameter `force.overwrite` to `findChromPeaks()` to allow overwriting of
  a HDF5 result file.
- Rename `.is_chrom_peaks_within_mz_rt()` to `.is_chrom_peak_within_mz_rt()`.

jorainer committed:
review: show supported values in error message

sneumann commented:
Excellent, thanks for the hard work and answering questions. Will merge next. Yours, Steffen
